### ISSUE TITLE
[question]/[feature request] Normalization for offline states [and rewards]

### ISSUE THREAD

**Issue by BruceK4t1qbit** [hill-a/stable-baselines#200]
Normalizing inputs generally accelerates training of machine learning models.  Does stable baselines have an approach for normalization in offline algorithms?  Or can we create one?

Using VecNormalize to normalize states/rewards would provide inconsistent normalization to samples in the replay buffer that were collected at different times.

**Comment by araffin** [comment]
Hello,

>Does stable baselines have an approach for normalization in offline algorithms?

You mean for normalizing in an offline manner? or for off-policy algorithms ?

If you have a constant normalization, can't you write it as part of your environment? (that is to say, the normalization is done before returning the observation and reward to the agent)
Or, if you want to be more general, you can create a custom VecEnv.

**Comment by BruceK4t1qbit** [comment]
This would be non-constant normalization for off-policy algorithms that use a replay buffer.  For real-world applications or where you can't rely on the environment pre-normalizing states/rewards reasonably.

I imagine this could be implemented using non-trainable variables in the tensorflow graph.

**Comment by araffin** [comment]
>This would be non-constant normalization for off-policy algorithms that use a replay buffer.

I don't get it. Could you elaborate? (because if you do a constant normalization in the env, I don't understand how it would not be constant for the agent)

>For real-world applications or where you can't rely on the environment pre-normalizing states/rewards reasonably.

It seems contradictory with what you want. On one hand, you say that you want a normalization that stay the same (constant) and on the other hand it seems that you want to learn it (for "real world applications").
Also, I think that inconsistency is only valid for very few samples, as the normalization usually converges quickly.
A solution for you, if you want to re-evaluate samples from the replay buffer with a new normalization, would be to store the raw observation in the buffer and apply current normalization when sampling them.

**Comment by BruceK4t1qbit** [comment]
To "store the raw observation in the buffer and apply current normalization when sampling them" is exactly what I'm interested in.  Is this implemented?


I've tried to lay out my point more clearly below by introducing some notation.

Let tci = t_c(sample_i) = the time that "sample_i" was collected
Let tb = denote the time when you are training on batch "b"

Let N_t(sample) denote normalization of "sample" WRT all data collected by time "t"

The problem with using VecNormalize (that normalizes as samples are produced by the environment) and replay buffers together is that, when training, samples in a single training batch are normalized differently/inconsistently.  So, for instance, at time tb you would train on a batch like:

N_tc1(sample_1)
N_tc2(sample_2)
...
N_tcNB(sample_NB)

This inconsistent normalization makes it impossible for a neural network to learn well.

What we want is, at time tb, to apply N_tb to all samples, so that normalization is consistent within each batch during training:

N_tb(sample_1)
N_tb(sample_2)
...
N_tb(sample_NB)

So the normalization has to be applied to samples each time a batch is sampled from the replay buffer for training, not prior to inserting each sample into the replay buffer...

**Comment by araffin** [comment]
>To "store the raw observation in the buffer and apply current normalization when sampling them" is exactly what I'm interested in. Is this implemented?

It is not implemented but should not be too much work for you:
1.  you will need to use `.get_original_obs()` from the [VecNormalize wrapper](https://stable-baselines.readthedocs.io/en/master/guide/vec_envs.html#stable_baselines.common.vec_env.VecNormalize.get_original_obs)
2. store this observation in the buffer (instead of the normalized one)
3. normalize when sampling from the buffer (calling `_normalize_observation()` method of the wrapper)

For that, you need to change directly the code of the algorithms (see install from source instructions).

>This inconsistent normalization makes it impossible for a neural network to learn well.

From my experience, I mostly disagree with this statement (the [rl zoo](https://github.com/araffin/rl-baselines-zoo) where several agents are successfully trained using this wrapper)

I don't think this will make a big difference but I'm still interested by your results ;) (we would consider integrating this mechanism if it helps the agent to learn)

**Comment by araffin** [comment]
On reflection, we should update VecNormalize (add `.get_original_reward()`) and the algos using the replay buffer (DQN, DDPG, SAC, TD3) especially for the normalization of the reward.
